signer/src/util.rs

```diff
@@ -0,0 +1,253 @@
+//! General utilities for the signer.
+
+use std::{
+    cmp::min,
+    future::Future,
+    ops::Deref,
+    sync::{
+        atomic::{AtomicU8, AtomicUsize, Ordering},
+        Arc,
+    },
+};
+
+use crate::error::Error;
+
+/// A fallback-wrapper that can failover to other clients if the current client fails.
+#[derive(Clone)]
+pub struct ApiFallbackClient<T> {
+    inner: Arc<InnerApiFallbackClient<T>>,
+}
+
+impl<T> Deref for ApiFallbackClient<T> {
+    type Target = InnerApiFallbackClient<T>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+/// Inner implementation of the fallback client.
+pub struct InnerApiFallbackClient<T> {
+    inner_clients: Vec<T>,
+    last_client_index: AtomicUsize,
+    retry_count: AtomicU8,
+}
+
+impl<T> InnerApiFallbackClient<T> {
+    /// Set the number of retries to perform before giving up.
+    pub fn set_retry_count(&self, retry_count: u8) {
+        self.retry_count.store(retry_count, Ordering::Relaxed);
+    }
+
+    /// Get a reference to the current inner API client.
+    pub fn get_client(&self) -> &T {
+        &self.inner_clients[self.last_client_index.load(Ordering::Relaxed)]
+    }
+
+    /// Execute a closure on the current client, falling back to remaining clients
+    /// if the closure returns an error.
+    pub async fn exec<'a, R, E, F>(&'a self, f: impl Fn(&'a T) -> F) -> Result<R, Error>
+    where
+        E: std::error::Error + std::fmt::Debug,
+        E: Into<Error>,
+        F: Future<Output = Result<R, E>> + 'a,
+    {
+        let retry_count = self.retry_count.load(Ordering::Relaxed);
+        for i in 0..retry_count {
+            let client_index = self.last_client_index.load(Ordering::Relaxed);
+            let result = f(&self.inner_clients[client_index]).await;
+
+            if let Err(error) = result {
+                tracing::warn!(%error, retry_num=i, max_retries=retry_count, "failover client call failed");
+                self.last_client_index.store(
+                    (client_index + 1) % self.inner_clients.len(),
+                    Ordering::Relaxed,
+                );
+                continue;
+            }
+
+            return result.map_err(Into::into);
+        }
+
+        Err(Error::AllFailoverClientsFailed)
+    }
+}
+
+impl<T> ApiFallbackClient<T> {
+    /// Create a new fallback client from a list of clients.
+    pub fn new(clients: Vec<T>) -> Self {
+        let retry_count = min(3, clients.len());
```

Comment:
Probably better to make the `retry_count` an explicit input or a global constant.

Reply:
I felt like this was a reasonable default -- if you give 5 endpoints you want it to try at least 5 times to touch all endpoints, and if you have one endpoint you probably want it to try at least 2-3 times. It can be updated using `set_retry_count()` otherwise..?

Reply:
Oh I was sloppy with my wording. I was trying to say that it's probably better to do something like
```rust
pub fn new(clients: Vec<T>, retry_count: usize) -> Self {
    ...
```
or
```rust
pub fn new(clients: Vec<T>, retry_count: Option<usize>) -> Self {
    let retry_count = retry_count.unwrap_or_else(|| clients.len());
```
or
```rust
/// blah blah blah
const DEFAULT_MINIMUM_RETRY_COUNT: usize = 3;

pub fn new(clients: Vec<T>) -> Self {
    let retry_count = min(DEFAULT_MINIMUM_RETRY_COUNT, clients.len());
```
or maybe some combination of the above.

Reply:
Yeah no I got it, I just didn't really want to force "users" to calculate their own retry counts when newing up (where they can instead use the `set_retry_count()` fn to set it manually _if they want to_). But I've at least changed it to a constant now so it's not buried in the code :+1: